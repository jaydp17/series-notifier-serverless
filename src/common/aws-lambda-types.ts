/**
 * Contains types that are related to AWS Lambda
 */

export type LambdaEvent = {
  httpMethod: string;
  body: string;
  queryStringParameters: {} | null;
};

/**
 * Callback type that should be used when the lambda is gonna be invoked with an HTTP req
 */
export type LambdaHttpCallback = (error: Error | null | undefined, response?: { statusCode: number, body?: string }) => void;

/**
 * Non-HTTP lambda callback type
 */
export type LambdaCallback = (error: Error | null | undefined, response?: {}) => void;
