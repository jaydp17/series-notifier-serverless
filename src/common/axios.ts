/**
 * Gives a common axios instance across the entire app
 */

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import Logger from '../common/logger';

const timeLogger = Logger('TimeLogger');

function requestInterceptor(config: AxiosRequestConfig) {
  return { ...config, startTime: Date.now() };
}

function responseInterceptor(response: AxiosResponse) {
  const { startTime } = response.config as any; // tslint:disable-line:no-any
  const endTime = Date.now();
  timeLogger.log(`[TimeLogger]:: ${response.config.url} -> ${endTime - startTime}ms`);
  return response;
}

axios.interceptors.request.use(requestInterceptor);
axios.interceptors.response.use(responseInterceptor);

export default axios;
