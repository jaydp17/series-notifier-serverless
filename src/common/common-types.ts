/**
 * This file contains types that are consumed by more than 1 function
 */

export type Callback = (error: Error | null | undefined, data?: any) => void; // tslint:disable-line:no-any
