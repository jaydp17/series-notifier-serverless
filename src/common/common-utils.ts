/**
 * Contains utility functions
 */

import { inspect } from 'util';

/**
 * Pretty prints an object
 */
export function prettyPrint(obj: any): void { // tslint:disable-line:no-any
  console.log(inspect(obj, false, 20, true)); // tslint:disable-line:no-console
}
