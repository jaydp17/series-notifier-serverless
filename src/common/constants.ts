/**
 * Contains constants used accross the app
 */
export const GRAPH_API_URL = 'https://graph.facebook.com/v2.6';
