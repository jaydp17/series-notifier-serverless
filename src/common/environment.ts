/**
 * Contains all the environment vailables
 */

/**
 * It's the NODE_ENV
 */
export const env: string = process.env.NODE_ENV || 'development';

export const isProd: boolean = env === 'production';

/**
 * The Stage which this function is running in
 */
export const stage: string = env === 'production' ? 'production' : 'dev';

/**
 * Name of the service, it's not gonna change
 */
export const serviceName: string = 'snbot';

export const pageToken = <string>process.env.FB_PAGE_TOKEN;
if (!pageToken && env !== 'test') {
  throw new Error('FB_PAGE_TOKEN not found in env vars');
}

export const verifyToken = <string>process.env.FB_VERIFY_TOKEN;
if (!verifyToken && env !== 'test') {
  throw new Error('FB_VERIFY_TOKEN not found in env vars');
}

export const botUsername = <string>process.env.FB_USERNAME;
if (!botUsername && env !== 'test') {
  throw new Error('FB_USERNAME not found in env vars');
}

export const tvdbApiKey = <string>process.env.TVDB_API_KEY;
if (!tvdbApiKey && env !== 'test') {
  throw new Error('TVDB_API_KEY not found in env vars');
}

export const traktApiKey = <string>process.env.TRAKT_API_KEY;
if (!traktApiKey && env !== 'test') {
  throw new Error('TRAKT_API_KEY not found in env vars');
}

export const witAIToken = <string>process.env.WITAI_TOKEN;
if (!witAIToken && env !== 'test') {
  throw new Error('WITAI_TOKEN not found in env vars');
}

export const fanArtApiKey = <string>process.env.FAN_ART_API_KEY;
if (!fanArtApiKey && env !== 'test') {
  throw new Error('FAN_ART_API_KEY not found in env vars');
}

export const theMovieDbApiKey = <string>process.env.THE_MOVIE_DB_API_KEY;
if (!theMovieDbApiKey && env !== 'test') {
  throw new Error('THE_MOVIE_DB_API_KEY not found in env vars');
}
