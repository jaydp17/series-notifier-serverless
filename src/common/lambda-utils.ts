/**
 * Contains utility functions related to AWS Lambda
 */

import { Lambda } from 'aws-sdk';

import { messengerReply, processQuery } from './function-names';
import { AnyAction, AnyReplyKind } from './internal-message-types';

const lambda = new Lambda();

export const invokeFunction = (
  functionName: string,
  payload: {},
  invocationType: string = 'Event',
): Promise<Lambda.InvocationResponse> => {
  return lambda
    .invoke({
      FunctionName: functionName,
      InvocationType: invocationType,
      Payload: JSON.stringify(payload),
    })
    .promise();
};

export const invokeProcessQuery = (data: AnyAction) => invokeFunction(processQuery, data);

/**
 * Invokes the function that replies FB Messenger back
 */
export function invokeMessengerReply(event: AnyReplyKind): Promise<Lambda.InvocationResponse> {
  return invokeFunction(messengerReply, event);
}
