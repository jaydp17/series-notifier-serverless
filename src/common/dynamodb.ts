/**
 * Creates a dynamodb instance
 */

import { DynamoDB } from 'aws-sdk';
import { env } from './environment';

let dynamoDbOptions = {};
if (env === 'development' && process.env.IS_LOCAL) {
  dynamoDbOptions = { endpoint: 'http://localhost:8000' }; //tslint:disable-line
}
export const dynamodb = new DynamoDB(dynamoDbOptions);
export const dynamoDocClient = new DynamoDB.DocumentClient(dynamoDbOptions);

export default {
  get: params => dynamoDocClient.get(params).promise(),
  put: params => dynamoDocClient.put(params).promise(),
  query: params => dynamoDocClient.query(params).promise(),
  delete: params => dynamoDocClient.delete(params).promise(),
};
