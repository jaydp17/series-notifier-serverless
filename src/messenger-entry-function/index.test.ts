/**
 * Test file for the entry function
 */

jest.mock('../common/lambda-utils');
jest.mock('../common/messenger.api');

// mocks
import { invokeProcessQuery } from '../common/lambda-utils';
import { messengerAuth } from '../common/messenger.api';

// types
import { LambdaEvent, LambdaHttpCallback } from '../common/aws-lambda-types';

// other imports
import deepFreeze from 'deep-freeze';
import { handler } from './index';

describe('Entry Function', () => {

  // tslint:disable-next-line:mocha-no-side-effect-code
  const baseEventBody = deepFreeze({
    object: 'page',
  });

  // tslint:disable-next-line:mocha-no-side-effect-code
  const baseEvent = deepFreeze({
    httpMethod: 'POST',
    body: JSON.stringify(baseEventBody),
  });

  beforeEach(() => {
    // clear mocks
    (<jest.Mock<{}>>invokeProcessQuery).mockClear();
    (<jest.Mock<{}>>messengerAuth).mockClear();
  });

  it('should call messengerAuth on GET request with queryStringParameters', () => {
    const event: LambdaEvent = {
      httpMethod: 'GET',
      queryStringParameters: {},
      body: '',
    };

    handler(event, {}, jest.fn());
    expect(messengerAuth).toBeCalled();
  });

  it('should return ping on GET req w/o queryStringParameters', () => {
    // prepare
    const event: LambdaEvent = {
      httpMethod: 'GET',
      queryStringParameters: null,
      body: '',
    };

    // execute
    const callback = jest.fn();
    handler(event, {}, callback);

    // test
    expect(callback).toBeCalled();
    expect(callback.mock.calls[0]).toMatchSnapshot();
    expect((<jest.Mock<{}>>messengerAuth).mock.calls.length).toEqual(0);
    expect((<jest.Mock<{}>>invokeProcessQuery).mock.calls.length).toEqual(0);
  });

  it('rejects non page Type events from FB Messenger', () => {
    // prepare
    const event: LambdaEvent = {
      ...baseEvent,
      body: JSON.stringify({ ...baseEventBody, object: 'nonPage' }),
      queryStringParameters: null,
    };

    // execute
    const callback = jest.fn();
    handler(event, {}, callback);

    // test
    expect(callback).toBeCalled();
    expect(callback.mock.calls[0]).toMatchSnapshot();
    expect((<jest.Mock<{}>>messengerAuth).mock.calls.length).toEqual(0);
    expect((<jest.Mock<{}>>invokeProcessQuery).mock.calls.length).toEqual(0);
  });

  it('returns 403 on non GET/POST requests', () => {
    // prepare
    const event: LambdaEvent = {
      httpMethod: 'PUT',
      queryStringParameters: null,
      body: '',
    };

    // execute
    const callback = jest.fn();
    handler(event, {}, callback);

    // test
    expect(callback).toBeCalled();
    expect(callback.mock.calls[0]).toMatchSnapshot();
    expect((<jest.Mock<{}>>messengerAuth).mock.calls.length).toEqual(0);
    expect((<jest.Mock<{}>>invokeProcessQuery).mock.calls.length).toEqual(0);
  });
});
