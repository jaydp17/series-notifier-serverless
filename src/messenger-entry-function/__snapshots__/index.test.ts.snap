// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entry Function rejects non page Type events from FB Messenger 1`] = `
Array [
  null,
  Object {
    "body": "Invalid Object Type",
    "statusCode": 403,
  },
]
`;

exports[`Entry Function returns 403 on non GET/POST requests 1`] = `
Array [
  null,
  Object {
    "body": "Invalid httpMethod: PUT",
    "statusCode": 403,
  },
]
`;

exports[`Entry Function should return ping on GET req w/o queryStringParameters 1`] = `
Array [
  null,
  Object {
    "body": "ping",
    "statusCode": 200,
  },
]
`;
